### PR TITLE
[SYCL][Graph] Makes command graph functions thread-safe

### PR DESCRIPTION
**Pull request by mfrancepillois**
Adds mutual exclusions to make graph functions thread-safe.
Adds debugging functions to help check equivalence of two graphs
Improves threading tests to ensure implementation thread-safe property.

Addresses Issue: #85

### CODE REVIEW

File: sycl/test-e2e/Graph/Threading/begin_end_recording.cpp [hunk 1]
```diff
@@ -59,6 +62,34 @@ int main() {
...
+
+  for (unsigned i = 0; i < NumThreads; ++i) {
+    queue MyQueue;
+    GraphRef.begin_recording(MyQueue);
+    run_kernels_usm(MyQueue, Size, PtrA, PtrB, PtrC);
+    GraphRef.end_recording(MyQueue);
+  }
+
+  auto GraphImpl = sycl::detail::getSyclObjImpl(Graph);
+  auto GraphRefImpl = sycl::detail::getSyclObjImpl(Graph);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Threading/begin_end_recording.cpp]: I don't think testing `getSyclObjImpl()` is something we should be doing in `test-e2e`. "e2e" stands for End To End, and is intended for testing code users can write (no other e2e tests use  `getSyclObjImpl()`) . Users won't have access to `graph_impl.hpp` from if they use a binary release or only have an install directory from a build.

I think `hasSimilarStructure` is important for testing though, but what I think this means is that we need to move the threading tests which use`getSyclObjImpl`  from `test-e2e` to the command-graph unittest - `unittests/Extension/CommandGraph.cpp`.

File: sycl/source/detail/graph_impl.cpp [hunk 2]
```diff
@@ -482,8 +518,10 @@ node modifiable_command_graph::addImpl(const std::vector<node> &Deps) {
 node modifiable_command_graph::addImpl(std::function<void(handler &)> CGF,
                                        const std::vector<node> &Deps) {
   std::vector<std::shared_ptr<detail::node_impl>> DepImpls;
-  for (auto &D : Deps) {
-    DepImpls.push_back(sycl::detail::getSyclObjImpl(D));
+  {
+    for (auto &D : Deps) {
+      DepImpls.push_back(sycl::detail::getSyclObjImpl(D));
+    }
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: Seems like an unnecessary change.

File: sycl/source/detail/graph_impl.cpp [hunk 3]
```diff
@@ -378,6 +412,8 @@ exec_graph_impl::enqueue(const std::shared_ptr<sycl::detail::queue_impl> &Queue,
     return NewEvent;
   });
 
+  std::lock_guard<std::recursive_mutex> Lock(MMutexGraphExec);
+
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: Do we need a recursive mutex on the graph? For `findRealDeps()` (and really most recursive functions or any public functions, for example if they are called from each other) it would probably be better to move the impl to a private function and just lock and call the private impl in the public function.

The nodes present a problem here I suppose but I'll discuss that in another comment.

File: sycl/source/detail/graph_impl.hpp [hunk 4]
```diff
@@ -43,6 +43,9 @@ class node_impl {
   /// Command group object which stores all args etc needed to enqueue the node
   std::unique_ptr<sycl::detail::CG> MCommandGroup;
 
+  /// Protects all the fields that can be changed by class' methods.
+  mutable std::recursive_mutex MMutexNode;
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.hpp]: I'm a bit concerned we're overcomplicating things here with mutexes on the nodes themselves. It is a problem that the nodes are shared between modifiable and executable graphs but I think it's preferable to solve that problem rather than use mutexes here. IMO, nodes should be modified only by the owning graph, so we need to be copying them into the executable graph, or some other solution that separates them. This should be fairly straightforward apart from rebuilding the successors/predecessors, but that's not an unsolvable problem.

However, adding that to this task would increase the scope too much. As a stop-gap could we just always lock the modifiable graph when we are dealing with shared data (nodes)? Using a `std::shared_mutex` for aquiring read/write locks.

For recursive functions same as my other comment we should be having private impls and just lock and call them from the public function. Any other private functions that call them should be then using the private impl to avoid the need for recursive mutexes.

File: sycl/source/detail/graph_impl.hpp [hunk 5]
```diff
@@ -336,6 +353,9 @@ class node_impl {
 /// Implementation details of command_graph<modifiable>.
 class graph_impl {
 public:
+  /// Protects all the fields that can be changed by class' methods.
+  mutable std::recursive_mutex MMutexGraph;
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  mutable std::recursive_mutex MMutex;
```

I don't think `Graph` is necessary in the name here, since it is the sole mutex member of the graph. Same for the other mutexes.

File: sycl/test-e2e/Graph/Threading/begin_end_recording.cpp [hunk 6]
```diff
@@ -1,7 +1,5 @@
-// REQUIRES: level_zero, gpu, TEMPORARY_DISABLED
-// Disabled as thread safety not yet implemented
-
-// RUN: %clangxx -pthread -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// REQUIRES: level_zero, gpu
+// RUN: %clangxx -I%S/../../../source -std=c++20 -pthread -fsycl -fsycl-targets=%{sycl_triple} %s -o %t.out
```
Comment (Bensuo) [inline sycl/test-e2e/Graph/Threading/begin_end_recording.cpp]: I don't think we should be requiring C++20 here until the compiler moves to it. There are some utility functions in the unit tests (see `sycl/unittests/thread_safety/ThreadUtils.h`) that might provide inspiration for an alternative to `std::barrier`.

Same for the other tests that require it.

**Review by Bensuo — CHANGES_REQUESTED** [review]

File: sycl/test-e2e/Graph/Threading/begin_end_recording.cpp (hunk 1 above)
Comment (EwanC) [reply sycl/test-e2e/Graph/Threading/begin_end_recording.cpp]: We chatted earlier about the unit tests emulating pi/ur, so any tests which require a device to run on not being suitable for moving. It looks like most of these threading tests don't actually submit an executable graph. `submit.cpp` does, but we're not checking the graph structure there anyway. `finalize.cpp` does too, but I think we could modify it to remove the `handler::ext_oneapi_graph()` call.

So I think it's worth trying to port these tests to the unittest framework and seeing how that goes.

File: sycl/source/detail/graph_impl.cpp [hunk 7]
```diff
@@ -346,6 +356,7 @@ void exec_graph_impl::createURCommandBuffers(sycl::device Device) {
 }
 
 exec_graph_impl::~exec_graph_impl() {
+  MMutex.lock();
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp]: Can we use RAII for some of these `lock()`/`unlock()` usage combinations so that the mutex gets automatically unlocked when a variable goes out of scope, something like [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard)

Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: std::lock_guard does not enable to do lock_shared and unlock_shared. I therefore prefer not to mix std::lock_guard with lock()/unlock() coding style in the same file.
A solution might be to code a custom  lock_guard class that manages shared_lock?

File: sycl/source/detail/graph_impl.hpp (hunk 4 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp]: I modified the locking policy to protect entry points instead of than endpoints.

File: sycl/source/detail/graph_impl.cpp (hunk 3 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: I modified the locking policy to protect entry points instead of than endpoints.

File: sycl/test-e2e/Graph/Threading/begin_end_recording.cpp (hunk 1 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/Threading/begin_end_recording.cpp]: I moved threading tests that don't require a device to run to unitests.
However, I prefer to keep "finalize.cpp" test in test-e2e to keep looking for memory leak on this test. 
A thread access issue was only revealed by the memory leak check of this test.

File: sycl/source/detail/graph_impl.cpp (hunk 7 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.cpp]: You can use `std::shared_lock` and `std::unique_lock` for this purpose. Quite common to do something like this in the class declaration:

```c++
using ReadLock = std::shared_lock<std::shared_mutex>;
using WriteLock = std::unique_lock<std::shared_mutex>;
```
Then you can use them to get an RAII style lock:

```c++
void foo(){
  ReadLock lock(MMutex);
  ...
}
```

File: sycl/source/handler.cpp [hunk 8]
```diff
@@ -370,8 +370,10 @@ event handler::finalize() {
     // If we have a subgraph node we don't want to actually execute this command
     // graph submission.
     if (!MSubgraphNode) {
+      MExecGraph->MMutex.lock();
       event GraphCompletionEvent =
           MExecGraph->enqueue(MQueue, std::move(CGData));
+      MExecGraph->MMutex.unlock();
```
Comment (Bensuo) [inline sycl/source/handler.cpp]: Shouldn't we just be locking inside `enqueue()`? Since this is only calling a single method locking outside of the call seems unnecessary here.

File: sycl/source/detail/graph_impl.cpp [hunk 9]
```diff
@@ -284,8 +292,10 @@ sycl::detail::pi::PiExtSyncPoint exec_graph_impl::enqueueNode(
       sycl::property_list{});
 
   std::vector<sycl::detail::pi::PiExtSyncPoint> Deps;
-  for (auto &N : Node->MPredecessors) {
-    findRealDeps(Deps, N.lock());
+  {
+    for (auto &N : Node->MPredecessors) {
+      findRealDeps(Deps, N.lock());
+    }
   }
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: Was was this extra scope added here?

File: sycl/source/detail/graph_impl.cpp [hunk 10]
```diff
@@ -256,9 +261,12 @@ sycl::detail::pi::PiExtSyncPoint exec_graph_impl::enqueueNodeDirect(
     sycl::detail::pi::PiExtCommandBuffer CommandBuffer,
     std::shared_ptr<node_impl> Node) {
   std::vector<sycl::detail::pi::PiExtSyncPoint> Deps;
-  for (auto &N : Node->MPredecessors) {
-    findRealDeps(Deps, N.lock());
+  {
+    for (auto &N : Node->MPredecessors) {
+      findRealDeps(Deps, N.lock());
+    }
   }
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: Another extra scope.

File: sycl/source/detail/graph_impl.cpp [hunk 11]
```diff
@@ -36,6 +37,8 @@ namespace {
 void connectToExitNodes(
     std::shared_ptr<node_impl> CurrentNode,
     const std::vector<std::shared_ptr<node_impl>> &NewInputs) {
+  // we need to lock the node mutex to ensure that NewInputs are actually
+  // registered as a successor to an exit node.
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: The node mutex has been removed right? This comment seems out of date.

File: sycl/source/detail/graph_impl.cpp [hunk 12]
```diff
@@ -75,6 +78,7 @@ bool checkForRequirement(sycl::detail::AccessorImplHost *Req,
 
 void exec_graph_impl::schedule() {
   if (MSchedule.empty()) {
+    // we lock the associated modifiable graph during the schedule
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: Should this comment have been removed?

File: sycl/source/detail/graph_impl.cpp [hunk 13]
```diff
@@ -566,7 +593,13 @@ executable_command_graph::executable_command_graph(
     const std::shared_ptr<detail::graph_impl> &Graph, const sycl::context &Ctx)
     : MTag(rand()),
       impl(std::make_shared<detail::exec_graph_impl>(Ctx, Graph)) {
+  // Graph and GraphExec is read and written in this scoe so we lock
+  // this graph with full priviledges.
+  Graph->MMutex.lock();
+  impl->MMutex.lock();
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]: I don't think you need to lock the `exec_graph_impl` in the constructor here because it's not possible for anything else to have access to it at this point.

File: sycl/source/detail/graph_impl.hpp [hunk 14]
```diff
@@ -313,6 +479,113 @@ class graph_impl {
     MInorderQueueMap[QueueWeakPtr] = Node;
   }
 
+  /// Prints the contents of the graph to a text file in DOT format
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.hpp]: Was this not included as part of #252?

Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp]: Yes. You are right. I fixed it

File: sycl/source/detail/graph_impl.cpp (hunk 13 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: Makes sense. I fixed it

File: sycl/source/handler.cpp (hunk 8 above)
Comment (mfrancepillois) [reply sycl/source/handler.cpp]: I moved Mutex lock/unlock into the enqueue method.

File: sycl/source/detail/graph_impl.cpp (hunk 7 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: I updated the implementation this way.

File: sycl/source/detail/graph_impl.cpp [hunk 15]
```diff
@@ -676,12 +683,17 @@ bool modifiable_command_graph::begin_recording(
   return QueueStateChanged;
 }
 
-bool modifiable_command_graph::end_recording() { return impl->clearQueues(); }
+bool modifiable_command_graph::end_recording() {
+  graph_impl::WriteLock Lock(impl->MMutex);
+  bool ret = impl->clearQueues();
+  return ret;
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]:
```suggestion
  return impl->clearQueues();
```
Doesn't seem like this change was necessary.

**Review by Bensuo — APPROVED** [review]
LGTM! Just one minor comment

File: sycl/unittests/Extensions/CommandGraph.cpp [hunk 16]
```diff
@@ -20,6 +21,62 @@
 using namespace sycl;
 using namespace sycl::ext::oneapi;
 
+namespace {
```
Comment (julianmi) [inline sycl/unittests/Extensions/CommandGraph.cpp:27]: I think a brief comment might be useful here.

File: sycl/unittests/Extensions/CommandGraph.cpp [hunk 17]
```diff
@@ -20,6 +21,62 @@
...
+    std::shared_ptr<sycl::ext::oneapi::experimental::detail::node_impl> Node) {
+  if (Node->MSuccessors.size() > 1)
+    return false;
+
+  for (const auto &Succ : Node->MSuccessors) {
+    return Succ->depthSearchCount();
+  }
+  return true;
+}
+
```
Comment (julianmi) [inline sycl/unittests/Extensions/CommandGraph.cpp:308]: We might want to comment these kernels as well and not the diamond dependency. We might also want to check if can reuse these in other unit tests.

**Review by julianmi — APPROVED** [review]
LGTM.

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 16 above)
Comment (mfrancepillois) [reply sycl/unittests/Extensions/CommandGraph.cpp:27]: I added some comments to this namespace.

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 17 above)
Comment (mfrancepillois) [reply sycl/unittests/Extensions/CommandGraph.cpp:308]: I added some comments to the new functions.
I haven't found any tests that can benefit from these functions that create diamond dependencies.

File: sycl/source/detail/graph_impl.hpp [hunk 18]
```diff
@@ -326,6 +327,16 @@ class node_impl {
     return true;
   }
 
+  /// Recusively computes the number of successor nodes
+  /// @return number of successor nodes
+  size_t depthSearchCount() {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]: could maybe be 
```suggestion
  size_t depthSearchCount() const {
```

File: sycl/source/detail/graph_impl.hpp [hunk 19]
```diff
@@ -582,6 +595,18 @@ class graph_impl {
   void makeEdge(std::shared_ptr<node_impl> Src,
                 std::shared_ptr<node_impl> Dest);
 
+  // Returns the number of nodes in the Graph
+  // @return Number of nodes in the Graph
+  size_t getNumberOfNodes() {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  size_t getNumberOfNodes() const {
```

File: sycl/source/detail/graph_impl.hpp [hunk 20]
```diff
@@ -582,6 +595,18 @@ class graph_impl {
...
+    size_t NumberOfNodes = 0;
+
+    for (std::shared_ptr<node_impl> Node : MRoots) {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]: Can capture `Node` by reference - `const auto &Node : MRoots`

File: sycl/unittests/Extensions/CommandGraph.cpp [hunk 21]
```diff
@@ -694,3 +760,175 @@ TEST_F(CommandGraphTest, MakeEdgeErrors) {
   // Re-check graph structure to make sure the graph state has not been modified
   CheckGraphStructure();
 }
+
+TEST_F(CommandGraphTest, MultiThreadsBeginEndRecording) {
```
Comment (EwanC) [inline sycl/unittests/Extensions/CommandGraph.cpp]: Rather than prefixing all these tests with `MultiThreads`, we can define a new gtest fixture class that derives from `CommandGraphTest` with some common functionality then create the tests based on that:

e.g

```cpp
class MultiThreadGraphTest : public CommandGraphTest {

 MultiThreadGraphTest() : CommandGraphTest(), NumThreads(std::thread::hardware_concurrency()), SyncPoint(NumThreads) {
   Threads.reserve(NumThreads);
}


unsigned NumThreads; 
Barrier SyncPoint;
std::vector<std::thread> Threads;
};

TEST_F(MultiThreadGraphTest, BeginEndRecording) {
```

File: sycl/test-e2e/Graph/Threading/submit.cpp [hunk 22]
```diff
@@ -12,6 +10,7 @@
 // ZE_DEBUG=4 testing capability.
 
 #include "../graph_common.hpp"
+#include <detail/graph_impl.hpp>
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Threading/submit.cpp]: I don't think we need this header

File: sycl/test-e2e/Graph/Threading/finalize.cpp [hunk 23]
```diff
@@ -12,9 +10,34 @@
 // ZE_DEBUG=4 testing capability.
 
 #include "../graph_common.hpp"
+#include <detail/graph_impl.hpp>
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Threading/finalize.cpp:13]: Why did we keep this test as a `test-e2e` test but move the other tests which use `graph_impl` to unittests? Is there a reason this can't be a unittest?

**Review by EwanC — APPROVED** [review]
Threading changes themselves look really good, have few other minor points

File: sycl/source/detail/graph_impl.hpp (hunk 18 above)
Comment (reble) [reply sycl/source/detail/graph_impl.hpp]: Looks like this is a public function that is doing a graph traversal but it's not thread-safe. Are we okay with this or am I missing something? Couldn't find find any use outside of the tests. Is that the only intended purpose, or are we assuming that it's always called under a reader lock? If so you might want to add a comment.

File: sycl/source/detail/graph_impl.hpp [hunk 24]
```diff
@@ -620,11 +645,21 @@ class graph_impl {
...
 };
 
 /// Class representing the implementation of command_graph<executable>.
 class exec_graph_impl {
 public:
+  using ReadLock = std::shared_lock<std::shared_mutex>;
+  using WriteLock = std::unique_lock<std::shared_mutex>;
+
+  /// Protects all the fields that can be changed by class' methods.
+  mutable std::shared_mutex MMutex;
```
Comment (reble) [inline sycl/source/detail/graph_impl.hpp:671]: Since each executable and modifiable graph has its own mutex, and there is a 1:n mapping of modifiable to executable graphs I see the potential for a deadlock. Because the order how we lock the graph becomes important. Do we have a test in place that would catch this?

File: sycl/source/handler.cpp [hunk 25]
```diff
@@ -1045,15 +1050,28 @@ void handler::ext_oneapi_graph(
...
     ParentGraph = MGraph;
   }
 
+  ext::oneapi::experimental::detail::graph_impl::WriteLock ParentLock;
   // If a parent graph is set that means we are adding or recording a subgraph
   if (ParentGraph) {
+    // ParentGraph is read and written in this scope so we lock this graph
+    // with full priviledges.
+    // We only lock for Record&Replay API because the graph has already been
+    // lock if this function was called from teh explicit API function add
```
Comment (reble) [inline sycl/source/handler.cpp]:
```suggestion
    // lock if this function was called from the explicit API function add
```

File: sycl/source/detail/graph_impl.hpp (hunk 18 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp]: The current implementation ensures that user entries, i.e. using the public graph API (sycl/include/sycl/ext/oneapi/experimental/graph.hpp), are thread safe (cannot trigger errors or generate invalid states). But "graph_impl" and "exec_graph_impl" are not guaranteed to be thread safe.

File: sycl/test-e2e/Graph/Threading/submit.cpp (hunk 22 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/Threading/submit.cpp]: You're right. I removed it.

File: sycl/source/detail/graph_impl.hpp (hunk 24 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp:671]: If I understand your comment correctly, you are asking if there is a potential risk for deadlock due to cross mutex locking due to multiple graph_exec referencing the same graph_impl, right?
Thread1:               Thread2:
A.lock()                  B.lock()
B.lock()                  A.lock()
If so, given that graph_impl does not reference graph_exec, the only case where it could append is:
Thread1:                                                                            Thread2:
                                                                                           graph_impl.begin_recording(queue)
graph_impl.add(ext_oneapi_graph(graph_exec))              queue.submit(ext_oneapi_graph(graph_exec))

However, since graph_exec is only locked with read permission (handle.cpp l.1054), this cannot create a deadlock.
Also, it is not allowed to use Explicit API and Record&Replay API at the same time for the same graph.

File: sycl/unittests/Extensions/CommandGraph.cpp (hunk 21 above)
Comment (mfrancepillois) [reply sycl/unittests/Extensions/CommandGraph.cpp]: Nice suggestion. Thanks.

File: sycl/test-e2e/Graph/Threading/finalize.cpp (hunk 23 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/Threading/finalize.cpp:13]: This tests submits the finalized graph to a queue. So we need a device for this test. 
And I prefer to not remove graph submission to test the thread safety property of "ext_oneapi_graph()".

Comment (EwanC) [reply sycl/test-e2e/Graph/Threading/finalize.cpp:13]: Makes sense 👍

File: sycl/source/detail/graph_impl.hpp (hunk 24 above)
Comment (reble) [reply sycl/source/detail/graph_impl.hpp:671]: Yes your understanding is correct. It might help to document somewhere that we'll have to take the locks always in the same order, e.g. modifiable first then executable. This could help reducing the risk of introducing a deadlock in the future.

**Review by reble — APPROVED** [review]
LGTM

File: sycl/source/detail/graph_impl.hpp (hunk 24 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp:671]: I added a comment.
